zebra/zebra_nhg.c

```diff
@@ -836,6 +836,9 @@ static void zebra_nhg_handle_install(struct nhg_hash_entry *nhe)
 
 	frr_each_safe(nhg_connected_tree, &nhe->nhg_dependents, rb_node_dep)
 		zebra_nhg_set_valid(rb_node_dep->nhe);
+
+	/* Schedule route to be processed/installed now */
+	rib_update(RIB_UPDATE_NH_INSTALL);
```

Comment:
Hmm, this looks like complete passes through the rib as nexthops gets installed (or groups change) ? that seems like a potential scaling issue. Would it make sense to queue up or otherwise identify the routes that actually need to wait for nexthop group installs to complete?

Reply:
I think managing that waiting queue would be more complex than you think. I started down that approach initially and hit a number of edge cases in rib processing code.

A couple issues were quickly apparent. What happens if this route has been changed or is no longer the preference? If we the queue routes, we have to loop over the queue and remove the item. What if the nexthops change? Do we loop over the queue again and remove the routes matching the old nexthop every time? We could use back pointers but that means every route addition/deletion and nexthop addition/deletion will have to remove/add pointers continually. It got a bit complex the more I thought about the edge handling that would need to be done to handle the simple case of making a route wait for a nexthop.

I think doing any kind of async waiting thing is going to have a performance hit of some kind. It seemed to me then, lets keep it as simple as possible at least for now and save cycles as simply as we can. Hence, why I added it to the rib_update system. We can guarantee a rib_update for nexthops will only ever be scheduled on the system once and it will loop through every route waiting for a nexthop and reprocess that node (in case its no longer the preferred route entry). With this method, yes we walk the rib table, but we guarantee at least if we just installed 200 nexthops at once, we should only end up walking the table once.

The worst case scenario from my analysis is really then (install nexthop -> rib_update -> install nexthop ->rib_update ....) But for that to happen we would need to sending nexthops to the dataplane REALLY slowly right? Are my assumptions correct there?

I do think that I could add a global boolean flag to know whether there actually is a route waiting to be processed. With that, then the thread event would only happen if we know for sure there exists at least one route waiting to be installed. And then set that flag to false after we finish processing the table.